File: test/features/pages/data/announcement_service_test.dart
import 'dart:convert';

import 'package:clashkingapp/core/services/api_service.dart';
import 'package:clashkingapp/features/pages/data/announcement_service.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';

void main() {
  test('loads a paginated archive of current and past posts', () async {
    final client = MockClient((request) async {
      expect(request.url.path, endsWith('/app/posts'));
      expect(request.url.queryParameters['offset'], '20');
      return http.Response(
        jsonEncode({
          'items': [
            {
              'id': 'past',
              'title': 'Previous update',
              'subtitle': 'Release notes',
              'status': 'expired',
              'published_at': '2026-06-01T10:00:00Z',
            },
          ],
          'has_more': true,
          'next_offset': 21,
        }),
        200,
        headers: {'content-type': 'application/json'},
      );
    });
    final service = AnnouncementService(apiService: ApiService(client: client));

    final page = await service.getPublishedPosts(offset: 20);

    expect(page.items.single.status, 'expired');
    expect(page.items.single.publishedAt, DateTime.utc(2026, 6, 1, 10));
    expect(page.hasMore, isTrue);
    expect(page.nextOffset, 21);
  });

  test('keeps the API home order so pinned posts stay first', () async {
    final client = MockClient(
      (request) async => http.Response(
        jsonEncode({
          'items': [
            {'id': 'pinned', 'title': 'Pinned', 'subtitle': 'First'},
            {'id': 'newer', 'title': 'Newer', 'subtitle': 'Second'},
          ],
        }),
        200,
        headers: {'content-type': 'application/json'},
      ),
    );
    final service = AnnouncementService(apiService: ApiService(client: client));

    final announcements = await service.getActiveAnnouncements();

    expect(announcements.map((item) => item.id), ['pinned', 'newer']);
  });

  test('loads the exact published post used by a push notification', () async {
    final client = MockClient((request) async {
      expect(request.url.path, endsWith('/app/announcements/post-1'));
      return http.Response(
        jsonEncode({
          'item': {
            'id': 'post-1',
            'title': 'Update',
            'subtitle': 'New features',
            'story_url': 'https://cdn.example.com/update.html',
            'body_blocks': [
              {'type': 'paragraph', 'text': 'Details'},
            ],
          },
        }),
        200,
        headers: {'content-type': 'application/json'},
      );
    });
    final service = AnnouncementService(apiService: ApiService(client: client));

    final announcement = await service.getAnnouncement('post-1');

    expect(announcement?.id, 'post-1');
    expect(announcement?.storyUrl, 'https://cdn.example.com/update.html');
    expect(announcement?.body, contains('Details'));
  });
}

File: expo/src/features/home/data/announcement-service.test.ts
import { AppAnnouncement } from './app-announcement';
import { AnnouncementService, announcementTarget } from './announcement-service';
import { createContractTestApi } from '../../../core/api/contract-api.testing';

const announcement = {
  id: 'a',
  version: '1',
  title: 'T',
  subtitle: 'S',
  body_blocks: [],
  presentation_type: 'article',
  show_on_home: true,
  pinned_on_home: false,
  status: 'live',
};
const apiFor = (fetchImplementation: typeof fetch) =>
  createContractTestApi({ baseUrl: 'https://api.test', fetchImplementation });

describe('announcements', () => {
  it('maps targets exactly', () => {
    expect(announcementTarget('ios')).toBe('ios');
    expect(announcementTarget('android')).toBe('android');
    expect(announcementTarget('web')).toBe('all');
  });

  it('preserves API-returned media URLs without a client CDN rewrite', async () => {
    const bannerUrl = 'https://api.clashk.ing/v2/media/announcement-hero.png';
    const storyUrl = 'https://api.clashk.ing/v2/media/announcement-story.html';
    const fetchImplementation = jest.fn().mockResolvedValue(
      new Response(
        JSON.stringify({
          item: {
            ...announcement,
            banner_image_url: bannerUrl,
            presentation_type: 'story',
            story_url: storyUrl,
          },
        }),
      ),
    );
    const service = new AnnouncementService(apiFor(fetchImplementation), 'ios', () => 'en');
    await expect(service.getAnnouncement('a')).resolves.toMatchObject({
      bannerImageUrl: bannerUrl,
      storyUrl,
    });
  });

  it('converts supported body blocks to the same safe document structure', () => {
    const item = AppAnnouncement.fromJson({
      id: 'a',
      title: 'Title',
      subtitle: 'Subtitle',
      banner_image_url: 'https://example.com/hero.png',
      body_blocks: [
        { type: 'heading', text: '<News>' },
        { type: 'paragraph', text: 'A & B' },
        { type: 'bullet_list', items: ['One', ''] },
        { type: 'image', url: 'javascript:alert(1)', caption: 'bad' },
      ],
    });
    expect(item.body).toContain('&lt;News&gt;');
    expect(item.body).toContain('A &amp; B');
    expect(item.body).toContain('<li>One</li>');
    expect(item.body).not.toContain('javascript:');
    expect(item.hasReadableBody).toBe(true);
  });

  it('keeps active fetch best-effort and uses language-only locale', async () => {
    const fetchImplementation = jest
      .fn()
      .mockResolvedValueOnce(
        new Response(JSON.stringify({ item: announcement, items: [announcement] })),
      )
      .mockRejectedValueOnce(new Error('offline'));
    const service = new AnnouncementService(apiFor(fetchImplementation), 'ios', () => 'en_GB');
    await expect(service.getActiveAnnouncements()).resolves.toHaveLength(1);
    const request = fetchImplementation.mock.calls[0]![0] as Request;
    expect(request.url).toBe('https://api.test/v2/app/announcements/active?target=ios&locale=en');
    expect(request.headers.has('authorization')).toBe(false);
    await expect(service.getActiveAnnouncements()).resolves.toEqual([]);
  });

  it('resolves an exact notification through the public announcement route', async () => {
    const fetchImplementation = jest.fn().mockResolvedValue(
      new Response(
        JSON.stringify({
          item: {
            ...announcement,
            id: 'target',
            title: 'Target',
            subtitle: 'Notification',
          },
        }),
      ),
    );
    const service = new AnnouncementService(apiFor(fetchImplementation), 'android', () => 'fr_CA');

    await expect(service.getAnnouncement(' target ')).resolves.toMatchObject({ id: 'target' });
    const request = fetchImplementation.mock.calls[0]![0] as Request;
    expect(request.url).toBe('https://api.test/v2/app/announcements/target?locale=fr');
    expect(request.headers.has('authorization')).toBe(false);
  });

  it('keeps exact notification lookup best-effort', async () => {
    const fetchImplementation = jest.fn().mockRejectedValue(new Error('missing'));
    const service = new AnnouncementService(apiFor(fetchImplementation), 'web', () => 'en');

    await expect(service.getAnnouncement('missing')).resolves.toBeNull();
    expect(fetchImplementation).toHaveBeenCalledTimes(1);
  });
});
